Draw latex diagrams for bloqs using QPIC

tanujkhattar opened the pull request:
Adds support to draw latex diagrams for Bloqs using `qpic`. 


Here are some pictures generated by 
<img width="1366" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/c174bfb1-3895-4e14-90b2-df533f7d4280">

<img width="1305" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/51e7748f-1a37-40fa-869a-62d015c98815">

<img width="1308" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/1b28c206-afaa-4808-8dee-9a934f416837">

<img width="1300" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/9923ed61-fc7a-44e0-b191-0df7498e8588">

Many other improvements can be made. For example we use a square box instead of left / right triangles (by setting `shape=>` or `shape=<` in the `G` gates) because elongated triangles look bad. The musical score diagrams added by @mpharrigan use an arrow shape of the form `=>` where the width of the horizontal part `=` scales with length of the text but the head of the arrow `>` stays constant. qpic doesn't have an out of the box option to do this but I think we can achieve something similar using `op="valid_tickz_code_to_draw_an_arrow"` where the `valid_tickz_code_to_draw_an_arrow` needs to be figured out. I spent sometime playing around but gave up. We can do this as a follow up

Another potential improvement is the logic in line manager to allocate lines to given Soquets. In many diagrams shown above, the joined lines are placed below the original split lines even when they are equivalent. 


cc @mpharrigan @ncrubin @fdmalone

mpharrigan commented on qualtran/drawing/_show_funcs.py:
```diff
@@ -84,3 +86,12 @@ def show_flame_graph(*bloqs: 'Bloq', **kwargs):
     IPython.display.display(IPython.display.SVG(svg_data))
+
+
+def show_bloq_via_qpic(bloq: 'Bloq', width: int = 1000, height: int = 400):
```
this should likely just be a different argument for `type` in `show_bloq`

mpharrigan replied:
@tanujkhattar ^

tanujkhattar replied:
I consider this to still be pretty experimental so wanted to keep it separate, but I can add it to `show_bloq` as well if you think it's fine.

mpharrigan replied:
well let's not make it the default yet, but the "spirit" of the show functions is to provide the simplest possible entry point to getting something to show up in a notebook. I think the `type` parameter is perfect for this case

mpharrigan commented on qualtran/bloqs/hubbard_model.ipynb:
revert since this notebook isn't actually changed

tanujkhattar replied:
done

mpharrigan commented on qualtran/bloqs/util_bloqs.py:
```diff
@@ -113,7 +113,7 @@ def wire_symbol(self, soq: 'Soquet') -> 'WireSymbol':
             text = f'[{", ".join(str(i) for i in soq.idx)}]'
             return directional_text_box(text, side=soq.reg.side)
-        return directional_text_box(' ', side=soq.reg.side)
+        return directional_text_box('S', side=soq.reg.side)
```
motivate this change? presumably the splitting action should be implied by the circuit topology

tanujkhattar replied:
Using `symbol.text` leads to an empty symbol which `qpic` doesn't like for labels of the gate. I can probably hack around if needed

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+
+"""Classes for drawing latex diagrams for bloqs with QPIC - https://github.com/qpic/qpic.
+
+QPIC is not a dependency of Qualtran and must be manually installed by users via
+`pip install qpic`.
```
why?

tanujkhattar replied:
License issues. QPIC has a [GPL-3.0 license](https://github.com/qpic/qpic#GPL-3.0-1-ov-file) license which is restricted but I confirmed with the open source team that having a dependency like this (where we don't explicit depend on the tool but can produce output the tool can consume) is fine.

mpharrigan replied:
do you want to make a note that you can install it with pip but this module will invoke the qpic executable via subprocessing?

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    from qualtran import Bloq, Connection, Signature
+
+
+def _wire_name_prefix_for_soq(soq: Soquet) -> str:
+    return soq.pretty()
```
docstring? do you anticipate this being something other than `soq.pretty()`?

tanujkhattar replied:
Maybe, I just wanted to keep it abstracted away so if we find a diagram where `soq.pretty` is not suitable and we want to use something else, we know where to make the change

mpharrigan commented on qualtran/bloqs/mcmt/and_bloq.py:
```diff
@@ -160,7 +160,7 @@ def add_my_tensors(
     def wire_symbol(self, soq: 'Soquet') -> 'WireSymbol':
         if soq.reg.name == 'target':
-            return directional_text_box('∧', side=soq.reg.side)
+            return directional_text_box('&', side=soq.reg.side)
```
motivate this change? can qpic not do unicode? document this restriction somewhere in the wire symbol world. (The previous symbol was the unicode character for "LOGICAL AND")

tanujkhattar replied:
reverted

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    replacements = {'&': r'\&', '_': r'\_', '<': r'$<$', '>': r'$>$', '⨁': r'$\oplus$'}
+    for key, val in replacements.items():
+        label = label.replace(key, val)
+    scale_prefix = r'\scalebox{' + str(scale) + r'}{' if scale else ''
```
`if scale is not None` or do you want scale of 0 to mean default?

mpharrigan replied:
unnecessary r'' strings here

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+        label = label.replace(key, val)
+    scale_prefix = r'\scalebox{' + str(scale) + r'}{' if scale else ''
+    scale_suffix = r'}' if scale else ''
+    return r'\textrm{' + scale_prefix + str(label) + scale_suffix + r'}'
```
you call `str` on label but are already calling `.replace` on it, so this duck typing won't add anything

tanujkhattar replied:
removed

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    return soq.pretty()
+
+
+def _format_label_text(label: str, scale: Optional[float] = None) -> str:
```
I'd personally find this more readable if there were two seperate code paths for scale vs no scale

```python
if scale is not None:
  return rf'{\scalebox{scale}{label}}
else:
  return rf'{label}
```

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    return r'\textrm{' + scale_prefix + str(label) + scale_suffix + r'}'
+
+
+def _soq_symbol_to_text(symbol: WireSymbol) -> str:
+    assert hasattr(symbol, 'text')
```
update type annotation to be a union of the texty symbols?

tanujkhattar replied:
`_soq_to_symb` returns a `WireSymbol` instead of an overloaded type based on soquet. Refactored and added an assertion to make mypy happy

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+def _soq_symbol_to_text(symbol: WireSymbol) -> str:
+    assert hasattr(symbol, 'text')
+    symbol_text = symbol.text.strip()
+    if not symbol_text:
+        symbol_text = 'SJ'
```
??

tanujkhattar replied:
removed

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+
+def _soq_symbol_to_text(symbol: WireSymbol) -> str:
+    assert hasattr(symbol, 'text')
+    symbol_text = symbol.text.strip()
```
why do we need to strip?

tanujkhattar replied:
one or more blank spaces are considered equivalent in the qpic code. Therefore, I want to remove any blankspaces before / after symbol_text (eg: `' '` for split / join) before processing.

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    symbol_text = symbol.text.strip()
+    if not symbol_text:
+        symbol_text = 'SJ'
+    return 5 + 5 * len(symbol_text)
```
explain these values? Does this actually work? Does qpic use monospaced fonts?

tanujkhattar replied:
Yes, it works. The values are empirical and the goal is to rescale the width of rectangle based on the length of the label

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+        self._alloc_wires_with_prefix: Dict[str, Set[int]] = defaultdict(set)
+
+    def _wire_name_tuple_for_soq(self, soq: Soquet) -> Tuple[str, int]:
+        prefix = _wire_name_prefix_for_soq(soq)
```
make this a (static) method instead of a free function?

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+            self._alloc_wire_for_soq(cxn.left)
+        self.soq_map[cxn.right] = self.soq_map[cxn.left]
+
+    def add_bloq(self, bloq: 'Bloq', pred: List['Connection'], succ: List['Connection']) -> None:
```
bloq isn't used

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+        return _format_label_text(self.soq_to_wirename(soq))
+
+
+class QpicCircuit:
+    """Builds data corresponding to the input specification of a QPIC diagram"""
```
can you document this class and its public methods or make this class or its methods private? document public attributes

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    """Builds data corresponding to the input specification of a QPIC diagram"""
+
+    def __init__(self, signature: 'Signature'):
+        self.soq_map = {}
```
what is soq_map? Why isn't this handled by the wire manager?

tanujkhattar replied:
wire manager is responsible to allocate / deallocate new wires. `soq_map` is used to track the lifetime of a wire. In Cirq, wires have names but we don't have an equivalent notion of "wires" here. 

soq_map essentially maps every soquet to the leftmost soquet which has a wire originating and it has not been split / joined / deallocated yet.

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+
+    Args:
+        bloq: Bloqs to draw the latex diagram for
+        file_path: If specified, the output is stored at the file.
```
"otherwise the data is returned"

tanujkhattar replied:
done

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+        for bloq. If specified, the output is written to `file_path`.
+    """
+    cbloq = bloq.as_composite_bloq()
+    qpic_circuit = QpicCircuit(cbloq.signature)
```
What do you think about removing all the business logic from the QpicCircuit constructor? i.e.: in `__init__` just initialize the member variables; then in this function you can call

```python
qpic_circuit.set_up_signature(cbloq.signature)
for binst, pred, succ in cbloq.iter_bloqnections():
  qpic_circuit.add_bloq(pred, succ)
```

tanujkhattar replied:
done

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    qpic_circuit = QpicCircuit(cbloq.signature)
+    for binst, pred, succ in cbloq.iter_bloqnections():
+        qpic_circuit.add_bloq(binst.bloq, pred, succ)
+    data = qpic_circuit.wires + qpic_circuit.gates
```
make a convenience property on QpicCircuit `.data` ?

tanujkhattar replied:
done

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+    r"""Invoke `qpic` script to generate output diagram of type qpic/tex/pdf/png.
+
+    Outputs one of the following files, based on `output_type`:
+     - base_file_path + '.qpic' - Copies the input qpic_file to output_file.
```
base_file_path isn't an argument

tanujkhattar replied:
updated

mpharrigan commented on qualtran/drawing/qpic_diagram.py:
```diff
@@ -0,0 +1,332 @@
+
+    Args:
+        qpic_file: Path to file containing input that should be passed to the `qpic` script.
+        output_file: Optional path to the output where generated diagram should be saved. Defaults to
+            qpic_file.with_suffix(f".{output_type}")
+        output_type: Format of the diagram generated using qpic. Supported output types are one of
+            ['qpic', 'tex', 'pdf', 'png']
```
what happens if output_file extension and output_type conflict?

tanujkhattar replied:
I use temporary files with the right extension as intermediate files to store results and move them to the output file when the right file is computed, so it doesn't matter what the extension of the output file is

mpharrigan commented:
The font size looks large compared to e.g. the size of control-circles. Is there a way to control the font size?

tanujkhattar commented:
> The font size looks large compared to e.g. the size of control-circles. Is there a way to control the font size?

Changed the `scale` argument to `0.8` by default so all fontsizes are now smaller by default. Also made a bunch of other improvements. I've updated the pictures in my original PR comment. PTAL

tanujkhattar commented:
@mpharrigan This is ready for another look.

mpharrigan commented:
The diagrams look really nice. I think the slightly smaller font  classes them up

mpharrigan reviewed (approved):
LGTM modulo merging the show function
